[REEF-491] Initial Vortex implementation

johnyangk opened the pull request:
![150723_architecture](https://cloud.githubusercontent.com/assets/4114572/8843561/a67b00e8-313c-11e5-922f-d034c3c1c9f4.png)

## How to test
* Unit test
  * In the directory `reef-vortex`, run `mvn clean install`
* Integration test
  * You can simply test using `reef-tests` as usual, the test is included in there

## Implementation
* Java ThreadPool-like user interface
* Distributed execution of user-submitted functions
* Handling of resource preemption events
* Simple Vortex Tasklet scheduling mechanisms
* Simple unit/integration tests to verify correctness

## Vortex-REEF component mapping
Here's a brief description of how some of main Vortex components map to those of REEF
* VortexManager
  * Maps to REEF Driver event handlers 
  * Acts as the brain of the Vortex runtime, handling events from different parts of the system, scheduling/executing user tasks, and keeping track of running executors/tasks
* LocalExecutor
  * Maps to REEF Context
  * Receives commands, executes them, and sends the results back to VortexManager


JIRA
https://issues.apache.org/jira/browse/REEF-491

bgchun commented on lang/java/reef-applications/reef-vortex/README.md:
```diff
@@ -0,0 +1,17 @@
+Vortex
+========================
+Vortex is a runtime for low-priority jobs to effectively cope 
```
This description is an important application of Vortex, but it does not define what Vortex is. Can you add descriptions about tasklet abstraction, queue, and the thread API and then talk about low-priority and high-priority jobs briefly? We don't need to talk much about the latter since we're in the process of exploring.

markusweimer replied:
Also, can we have a summary of this text in the JavaDoc for the package? That way, it becomes discoverable on our website.

bgchun commented on lang/java/reef-applications/reef-vortex/README.md:
```diff
@@ -0,0 +1,17 @@
+How to use
+====================
+
+* Implement the `VortexUserCode` interface by executing tasks via `VortexThreadPool`
```
Code examples would be great. I like how the `Tang` documentation is written.

bgchun commented on lang/java/reef-applications/reef-vortex/pom.xml:
```diff
@@ -28,4 +28,140 @@ under the License.
         <version>0.12.0-incubating-SNAPSHOT</version>
     </parent>
+
+    <dependencies>
```
I cannot pinpoint, but I'm wondering if we can shorten this `pom.xml`.

bgchun replied:
`reef-vortex` or `vortex`?

johnyangk replied:
Changed to `vortex`, following how wake(and tang) is just `wake`, not `reef-wake`.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/VortexLauncher.java:
```diff
@@ -0,0 +1,85 @@
+                                           final int numOfExecutors,
+                                           final int executorMemory,
+                                           final int executorCores) {
+    final Configuration runtimeConfiguration = LocalRuntimeConfiguration.CONF
```
Inconsistent naming: `runtimeConfiguration` and `vortexConf`
`runtimeConfiguration` and `vortexConfiguration` OR
`runtimeConf` and `vortexConf`

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/VortexLauncher.java:
```diff
@@ -0,0 +1,85 @@
+    final Configuration runtimeConfiguration = LocalRuntimeConfiguration.CONF
+        .set(LocalRuntimeConfiguration.MAX_NUMBER_OF_EVALUATORS, MAX_NUMBER_OF_EVALUATORS)
+        .build();
+    final Configuration vortexConf =
```
`vortexConf` or `vortexDriverConf`?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFunction.java:
```diff
@@ -0,0 +1,35 @@
+ * @param <TInput> input type
+ * @param <TOutput> output type
+ */
+public interface VortexFunction<TInput extends Serializable, TOutput extends Serializable> extends Serializable {
```
We should avoid using `java.io.Serializable` since it's slow. Let's create a JIRA issue for improvement.

markusweimer replied:
+1 in principle. However, aren't they serializing code here? Which has its dangers in its own right (hence the creation of Tang), but I am not aware of another way to serialize code in Java.

johnyangk replied:
One way I can think of to avoid `java.io.Serializable` is to pre-load all user-defined functions at the time of Evaluator launch. That way Driver simply tells the class name of the function to Evaluator instead of serializing and sending the actual class. This also lets us avoid the redundant serialization/sending of same function classes.

I'm not sure how we can use Tang(pre-inject?) to make this happen and of the implications for the Vortex user api(need to let Vortex know all function classes the user will use before launching any tasks?) though.

johnyangk replied:
Created a JIRA for serialization:
https://issues.apache.org/jira/browse/REEF-504
We can maybe continue the discussion here?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFunction.java:
```diff
@@ -0,0 +1,35 @@
+  /**
+   * This will be called by a TaskletExecutor.
+   */
+  TOutput call(TInput input) throws Exception;
```
Are you asking users throw a checked exception when needed?

markusweimer replied:
Please document the exception with `@throws`. Specifically: When is an implementing class allowed to throw, and what happens if it does. A particularly fun edge case is: What if it throws, but has launched threads before that are still active? That case is a good candidate for "undefined behavior" :-)

johnyangk replied:
My idea is to allow users to throw whatever exceptions they want to throw in their `VortexFunction` implementations, catch them in `VortexWorker` to report to `VortexMaster`, and surface them to the user via `ExecutionException` in VortexFuture#get:

```java
public TOutput get() throws InterruptedException, ExecutionException
```

@bgchun I guess we allow them to throw any exception(unchecked+checked) they want, to be caught by `VortexWorker`

@markusweimer Am I understanding correctly that you are referring to a case where the function(thread A) spawns a new thread(thread B) on its own so when thread A throws, B gets orphaned? 

I guess the possibility of arbitrary resource usage by user code is also related with what @zpqian commented before: restricting the amount of resource a tasklet can use. Addressing this can be another new JIRA issue.

johnyangk replied:
Added a path to catch user thrown exceptions in `VortexWorker` and bubble them up all the way up to `VortexFuture#get`.

Added the following documentation:
```java
  /**
   * @param input of the function
   * @return output of the function
   * @throws Exception thrown here will bubble up in VortexFuture#get as ExecutionException
   * Exception should be thrown only after all resources are released as they cannot be cleared by Vortex
   * For example if threads are spawned here, shut them down before throwing an exception
   */
  TOutput call(TInput input) throws Exception;
```

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFuture.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * The interface between user code and submitted task.
+ */
+public class VortexFuture<TOutput> implements Future, Serializable {
```
Again, `Serializable` is not fast.
Why do you need `VortexFuture` implements `Serializable`?

bgchun replied:
The basic `Future` implementation's blocking. But I often found that having `Callback` (or call `Listener`) is useful.
That is when the task is completed, the registered user's `Callback` is called. So, I tend to create my own `Future` with callback support. Let's create a JIRA issue to track this improvement.

johnyangk replied:
My mistake, we don't need `Serializable` here.
Okay, I'll post a JIRA for the callback feature.

johnyangk replied:
New JIRA: Callback features for VortexFuture
https://issues.apache.org/jira/browse/REEF-505

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexThreadPool.java:
```diff
@@ -0,0 +1,47 @@
+
+  public <TInput extends Serializable, TOutput extends Serializable> VortexFuture<TOutput>
+      submit(final VortexFunction<TInput, TOutput> function, TInput input) {
+    // todo: Support duplicate tasklets / tasklet priorities.
```
todo -> TODO.
Another JIRA issue.

markusweimer replied:
We also have a syntax for todos:

```
// TODO[REEF-XXX]
```

You have to reference a JIRA with todos, as we otherwise will forget about them...

johnyangk replied:
Modified TODOs to point to respective JIRAs

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexThreadPool.java:
```diff
@@ -0,0 +1,47 @@
+      submit(final VortexFunction<TInput, TOutput> function, TInput input) {
+    // todo: Support duplicate tasklets / tasklet priorities.
+    final VortexFuture<TOutput> vortexFuture = new VortexFuture<>();
+    vortexManager.enqueueNewTasklet(new Tasklet<>(taskletId.getAndIncrement(), function, input, vortexFuture));
```
This interface's a bit strange.
`VortexManager` receives `Tasklet`s. You create a `VortexFuture` instance and pass it to the `Tasklet` instance. There's no return value from `enqueueNewTasklet`. 

I expected that `enqueueNewTasklet` returns a `VortexFuture` object which is created internally in the `enqueueNewTasklet` method.

johnyangk replied:
Changed to what you've suggested.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexUserCode.java:
```diff
@@ -0,0 +1,31 @@
+   * Your implementation of this interface will be instantiated by Tang
+   * and this method will be called upon the start of REEF Driver.
+   */
+  void start(final VortexThreadPool vortexThreadPool);
```
This is confusing. 
Why do I need `vortexThreadPool` as an argument in the start method? 
I expected to have `run` without any parameter.

What's the relationship between `VortexFunction` and `VortexUserCode`? 
Does user write `VortexUserCode`? This is confusing.

johnyangk replied:
@zpqian and I wanted to make it explicit for the Vortex users to use `VortexThreadPool` injected by Tang. Currently we inject `VortexThreadPool` in Vortex internal code and pass it to `VortexuserCode` via this method.

Then do you prefer the user by herself inject `VortexThreadPool` into the constructor of her implementation of `VortexUserCode` and use it inside `start()`? Or maybe should we somehow rethink the overall api design?

@zpqian, what are your thoughts?

zpqian replied:
I guess if we only allow a single VortexThreadPool instance, then the current design seems reasonable, or maybe we could use static methods or singleton instead?

Another possibility is to allow dynamic creation of VortexThreadPool instances, each of which manages a disjoint set of Executors. And in that case, we might just let user to do whatever he wants; but I'm not sure if this is something easy to support in a REEF Driver.

Zhengping, from my phone

> On Jul 24, 2015, at 8:15 AM, John Yang <notifications@github.com> wrote:
> 
> In lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexUserCode.java:
> 
> > + * KIND, either express or implied.  See the License for the
> > + * specific language governing permissions and limitations
> > + * under the License.
> > + */
> > +package org.apache.reef.vortex.api;
> > +
> > +/**
> > + * The starting point of Vortex user thread.
> > + */
> > +public interface VortexUserCode {
> > +  /**
> > +   * Implement this method, using VortexThreadPool to run jobs on Vortex.
> > +   * Your implementation of this interface will be instantiated by Tang
> > +   * and this method will be called upon the start of REEF Driver.
> > +   */
> > +  void start(final VortexThreadPool vortexThreadPool);
> @zpqian and I wanted to make it explicit for the Vortex users to use VortexThreadPool injected by Tang. Currently we inject VortexThreadPool in Vortex internal code and pass it to VortexuserCode via this method.
> 
> Then do you prefer the user by herself inject VortexThreadPool into the constructor of her implementation of VortexUserCode and use it inside start()? Or maybe should we somehow rethink the overall api design?
> 
> @zpqian, what are your thoughts?
> 
> —
> Reply to this email directly or view it on GitHub.
>

johnyangk replied:
As per the conference call today, we'll stick with the current approach. Let's revisit this later when we have more concrete Vortex applications.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/Command.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Initial version of Driver -> Evaluator protocol.
+ */
+public interface Command extends Serializable {
```
Again, this uses `Serializable`. If you think that this object is exchanged infrequently and its size is small, then `Serializable` is okay.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/Command.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public interface Command extends Serializable {
+  /**
+   * Type of Command.
```
What other types do you expect in the future?

zpqian replied:
(Duplicate) tasklet cancellation?

johnyangk replied:
Possible new types
* User cancellation of tasklet(s): https://issues.apache.org/jira/browse/REEF-502
* Duplicate tasklet cancellation: https://issues.apache.org/jira/browse/REEF-500

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/CommandExecuteTasklet.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class CommandExecuteTasklet<TInput extends Serializable, TOutput extends Serializable> implements Command {
+  private final int taskletId;
+  // todo: tasklet priority
```
todo -> TODO

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/Report.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Initial version of Evaluator -> Driver protocol.
+ */
+public interface Report extends Serializable {
```
`Serializable`

`Report` seems to be a generic name.

johnyangk replied:
Changed to `WorkerReport` for consistency. 
(Throughout the codebase renamed classes/variables to follow VortexMaster-VortexWorker structure)

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/CommandSender.java:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Takes the serialization work from the scheduler thread.
+ * todo(not urgent): See if we can make changes in REEF internals to
```
This TODO is not comprehensible.

johnyangk replied:
Since this is actually a minor issue, I'll remove it altogether.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/CommandSender.java:
```diff
@@ -0,0 +1,50 @@
+ * todo(not urgent): See if we can make changes in REEF internals to
+ * have the RemoteManager Thread pool do this work.
+ */
+public class CommandSender {
```
Why `CommandSender`? Why not `TaskletSender` if we ship `Tasklet`.

zpqian replied:
The purpose of the additional wrapper is to allow some sort of protocol between a Driver and the Executors. One potential use is to cancel a (duplicate) tasklet; another is to separate (input) data from code (i.e., VortexFunction) shipping in case multiple tasklets share a same input.

johnyangk replied:
Changed the method name of this class from `launch` to `send` to make it less confusing.

johnyangk replied:
As per the conference call today, changed the name to `VortexRequestor`.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexManager.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.logging.Logger;
+
+/**
+ * Default implementation of VortexManager.
```
Improve this comment since `VortexManager` is a key component.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexManager.java:
```diff
@@ -0,0 +1,92 @@
+  private static final Logger LOG = Logger.getLogger(DefaultVortexManager.class.getName());
+  private final RunningExecutors runningExecutors;
+  private final BlockingDeque<Tasklet> pendingTasklets = new LinkedBlockingDeque<>();
+  private final ExecutorService schedulerThread = Executors.newSingleThreadExecutor();
```
I'm wondering if we'd like to use `wake` `Stage`s instead of `Executors` directly.

johnyangk replied:
I personally prefer the current `ExecutorService`
* This thread directly accesses pendingTasklets/runningExecutors, which currently are directly accessed only by methods in this class. It maybe better to keep the accesses all in the same class for readability.
* I'm not sure how we can map execution of an infinite loop to the `onNext(event)` interface well.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexManager.java:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  @Override
+  public void enqueueNewTasklet(final Tasklet tasklet) {
```
I was thinking `VortexManager`s can be accessed concurrently by multiple threads. Is this not the case? 
If it's concurrently accessed, this class should be updated to be thread-safe.

johnyangk replied:
Yes it is accessed concurrently by multiple threads(the user thread, the schedulet thread and `EvaluatorMessageDispatcher` threads). Concurrency control is done inside 2 data structures(`pendingTasklets` and `runningExecutors`) it uses, so we do not have any locks or synchronized blocks in this class. 

I'll add this to the documentation.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexManager.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Default implementation of VortexManager.
+ */
+public class DefaultVortexManager implements VortexManager {
```
It looks like we need a method to cancel a `Tasklet`. It is still in `pendingTasklets`, but we don't need to execute it any more.

zpqian replied:
Great point! I guess one way to do it, is to add method for cancellation in VortexFuture.

johnyangk replied:
Would it be okay if we address this separately later? The related methods in `VortexFuture` are currently blocked.

```java
  @Override
  public boolean cancel(final boolean mayInterruptIfRunning) {
    throw new UnsupportedOperationException("Cancel not yet supported");
  }

  @Override
  public boolean isCancelled() {
    throw new UnsupportedOperationException("Cancel not yet supported");
  }

```

johnyangk replied:
Created a JIRA for this:
https://issues.apache.org/jira/browse/REEF-502

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+ * REEF Driver for Vortex.
+ */
+@Unit
+public final class VortexDriver {
```
Could you add proper annotations (@DriverSide, @TaskSide) throughout the code base?

johnyangk replied:
Sure I'll add them, although I'm not really sure `@TaskSide` is a right description for Vortex which doesn't really make use of `REEF Task`.

johnyangk replied:
As per the conference call today, I'll make use of `REEF Task`.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+@Unit
+public final class VortexDriver {
+  private static final Logger LOG = Logger.getLogger(VortexDriver.class.getName());
+  private final EvaluatorRequestor evaluatorRequestor;
```
Please add comments to private members in this class.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+  private final int evalCores;
+
+  private final VortexThreadPool vortexThreadPool;
+  private final ExecutorService vortexUserThread = Executors.newSingleThreadExecutor();
```
Encoding a single thread is not good. We should make it configurable, although you can use one as a default value. 
Also, the question about using `wake` `Stage` still holds.

johnyangk replied:
Agreed. Created `VortexStartExecutor implements EventHandler<VortexStart>` and moved the related code to there. This event handler is executed by a wake `ThreadPoolStage` whose number of threads is configurable. 

But, as the current code assumes single thread pool, I did not expose configuration to `VortexLauncher`. We can expose this parameter when we work on supporting multiple `VortexStart` on a single REEF Driver run later.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+
+      // Run the user code as a separate thread
+      // todo: handle user code exceptions
+      vortexUserThread.execute(new Runnable() {
```
Can we think of a better pattern?

johnyangk replied:
Yes, now we call
```java
vortexStartEStage.onNext(vortexStart);
```
in the constructor. The code is moved out of `VortexDriver` to there.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+  private final CommandSender commandSender;
+
+  private final int evalMem;
+  private final int evalNum;
```
The initial evalNum is fixed. But the actual number of evaluators can change over time due to elasticity. We should be able to maintain this information. 
Do we target homogeneous evaluators? This is a good start. I'm just checking if we need to support heterogeneous evaluators for the first target Vortex application.

zpqian replied:
Alternatively, does it make sense to have Executors report back (or update on the fly) available resources?

johnyangk replied:
Yes, initially we assume single distributed thread pool with homogeneous evaluators. I think runtime resource configuration/monitoring can be separate JIRA issue(s).

johnyangk replied:
Registered a JIRA:
[Extend configuration options for Vortex resources](https://issues.apache.org/jira/browse/REEF-517)

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+
+      final Configuration contextConfiguration = ContextConfiguration.CONF
+              .set(ContextConfiguration.IDENTIFIER, allocatedEvaluator.getId() + "_context")
+              .set(ContextConfiguration.ON_MESSAGE, LocalExecutor.class)
```
LocalExecutor?

johnyangk replied:
I agree it is a bad name. I'm thinking of renaming it to `VortexWorker`. (And `VortexManager` to `VortexMaster` for symmetricity)

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+      final Configuration contextConfiguration = ContextConfiguration.CONF
+              .set(ContextConfiguration.IDENTIFIER, allocatedEvaluator.getId() + "_context")
+              .set(ContextConfiguration.ON_MESSAGE, LocalExecutor.class)
+              .set(ContextConfiguration.ON_SEND_MESSAGE, LocalExecutor.class)
```
Same comment, here.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+    @Override
+    public void onNext(final ActiveContext activeContext) {
+      LOG.log(Level.INFO, "Evaluator up and running");
+      vortexManager.taskletExecutorAllocated(new DefaultTaskletExecutor(commandSender, activeContext));
```
Yep. `VortexManager` should be thread safe.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexDriver.java:
```diff
@@ -0,0 +1,180 @@
+  /**
+   * Evaluator preempted.
+   */
+  public final class FailedEvaluatorHandler implements EventHandler<FailedEvaluator> {
```
We need to implement `FailedTaskHandler`, too. Right?

johnyangk replied:
I don't think so. We do not use `REEF Task`. 

We may want to implement `FailedContextHandler` later though, when we want to distinguish resource preemptions by RM from other types of failures as we've discussed. I'll create a JIRA issue for that.

johnyangk replied:
New JIRA for this:
Distinguish different types of FailedEvaluator in Vortex
https://issues.apache.org/jira/browse/REEF-501

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexManager.java:
```diff
@@ -0,0 +1,57 @@
+ * Processes various tasklet related events coming from different components of the system.
+ */
+@DefaultImplementation(DefaultVortexManager.class)
+public interface VortexManager {
```
Please refer to the comments I left at the `VortexManager` implementation. It'd be great to crystallize `VortexManager`, `VortexFunction`, `Tasklet`, `VortexUserCode`, `VortexThreadPool`, etc.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/scheduler/RandomTaskletScheduler.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  @Override
+  public TaskletExecutor schedule(final Collection<TaskletExecutor> availableExecutors, final Tasklet tasklet) {
```
Passing `Collection` does not make sense.
You need to do complicated operations with `availableExecutors`. 
For example, if you can do lookup by index, you don't need to iterate over the collection.

johnyangk replied:
Yes, I agree. I'll think of a better structure for this.

johnyangk replied:
Since we are still in the stage of exploring ways to optimize the data structure for scheduling, for now I will just remove the scheduler package altogether and embed the random scheduler into `RunningExecutors` class

Once we have better understanding of key metrics(e.g. resource reliability, queue size, data locality) for making scheduling decisions we can revisit this as a separate JIRA issue. For example we can maintain data structures in `RunningExecutors` in a way that enables fast scheduling using indices on those metrics.

johnyangk replied:
Registered a JIRA for this:
[Smarter VortexWorker-Tasklet scheduling](https://issues.apache.org/jira/browse/REEF-518)

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/scheduler/TaskletScheduler.java:
```diff
@@ -0,0 +1,38 @@
+   * One can configure and experiment with different scheduling policies
+   * by implementing different flavors of this method.
+   */
+  TaskletExecutor schedule(final Collection<TaskletExecutor> availableExecutors, final Tasklet tasklet);
```
Do you have to pass availableExecutors? 
Do you think we can pass it when we construct a `TaskletScheduler` instance? 
Do you expect to have a set of `availableExecutors`?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/DefaultTaskletExecutor.java:
```diff
@@ -0,0 +1,79 @@
+  @Override
+  public <TInput extends Serializable, TOutput extends Serializable>
+      void execute(final Tasklet<TInput, TOutput> tasklet) {
+    runningTasklets.put(tasklet.getId(), tasklet);
```
What do we do when there exists a tasklet with the same id? Probably, raise an exception?

bgchun replied:
Do we always guarantee that the id is unique?

johnyangk replied:
Yes we do guarantee uniqueness of id, since it purpose is to distinguish a tasklet from others. I'll see if we can add some checkers to throw an exception if this constraint is founded to be violated.

johnyangk replied:
Added the following line right before this line:
```java
    assert(!runningTasklets.containsKey(tasklet.getId()));
```

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/DefaultTaskletExecutor.java:
```diff
@@ -0,0 +1,79 @@
+
+  @Override
+  public <TOutput extends Serializable>
+      void complete(final Integer taskletId, final TOutput result) {
```
Who's calling this `complete`?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/DefaultTaskletExecutor.java:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  @Override
+  public Collection<Tasklet> getTasklets() {
```
Where do you need `getTasklets`?

johnyangk replied:
Thanks for the catch. This method was actually being used for both the internal code and unit tests which is not good. 

I renamed it to `preempted` for the internal use, and created a new `containsTasklet` method for the unit tests to use.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+ * Keeps track of all running TaskletExecutors and Tasklets.
+ * todo: create unit tests to test possible race conditions
+ */
+public class RunningExecutors {
```
We're missing profiling here and there (e.g., number of pending tasklets over time, etc.). Let's create a JIRA issue for that.

bgchun replied:
Annotation.

bgchun replied:
Can you add unit tests for this class?

johnyangk replied:
Okay, I will create new unit tests for testing the possible race conditions I wrote in the diagram.

johnyangk replied:
New JIRA: Basic Vortex profiling
https://issues.apache.org/jira/browse/REEF-503

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+  private final TaskletScheduler taskletScheduler;
+
+  // RunningExecutors and its locks
+  private final HashMap<String, TaskletExecutor> runningExecutors = new HashMap<>();
```
Can you add comments to these private members?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+   * Parameter: Called exactly once per taskletExecutor.
+   */
+  public void addExecutor(final TaskletExecutor taskletExecutor) {
+    if (!terminated) {
```
Add detailed descriptions to this block.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+   * Parameter: Called exactly once per id.
+   */
+  public Collection<Tasklet> removeExecutor(final String id) {
+    if (!terminated) {
```
Same here

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+   * Parameter: Same tasklet can be launched multiple times.
+   */
+  public void launchTasklet(final Tasklet tasklet) {
+    if (!terminated) {
```
Same here

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/RunningExecutors.java:
```diff
@@ -0,0 +1,215 @@
+  public void completeTasklet(final String taskletExecutorId,
+                              final int taskletId,
+                              final Serializable result) {
+    if (!terminated) {
```
Same here

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/Tasklet.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Representation of user task in Driver.
+ */
+public class Tasklet<TInput extends Serializable, TOutput extends Serializable> implements Serializable {
```
Annotation

bgchun replied:
Please refer to the comments I left above.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/TaskletExecutor.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Representation of a TaskletExecutor in Driver.
+ */
+public interface TaskletExecutor {
```
Do we maintain multiple `TaskletExecutor` objects?

johnyangk replied:
Removed the interface and just made DefaultTaskletExecutor  -> TaskletExecutor as this is something @zpqian also pointed out before.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/tasklet/TaskletExecutor.java:
```diff
@@ -0,0 +1,46 @@
+   */
+  <TInput extends Serializable, TOutput extends Serializable> void execute(final Tasklet<TInput, TOutput> tasklet);
+
+  <TOutput extends Serializable> void complete(final Integer taskletId, final TOutput result);
```
`complete` or `completed`?

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/LocalExecutor.java:
```diff
@@ -0,0 +1,115 @@
+    final ExecutorService executor = Executors.newFixedThreadPool(evalCores); // number of threads = specified evalCores
+    schedulerThread.execute(new Runnable() {
+      @Override
+      public void run() {
```
Need more step-by-step descriptions.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addone/AddOneUserCode.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * AddOne User Code Example.
+ */
+public class AddOneUserCode implements VortexUserCode {
```
Okay. I'm less confused. We need a better name. `VortexUserCode` does not work well.

johnyangk replied:
Changed to `VortexStart`. Hope this makes more sense.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addone/AddOneUserCode.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public void start(final VortexThreadPool vortexThreadPool) {
```
Passing `vortexThreadPool` is strange.

bgchun commented on lang/java/reef-applications/reef-vortex/src/test/java/org/apache/reef/vortex/test/DefaultVortexManagerTest.java:
```diff
@@ -0,0 +1,138 @@
+  private TestUtil testUtil = new TestUtil();
+
+  @Test
+  public void testSingleTaskletNoFailure() throws Exception {
```
Add comments to the tests. For each test, give a concise description on what you're testing.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addone/AddOneFunction.java:
```diff
@@ -0,0 +1,28 @@
+
+import org.apache.reef.vortex.api.VortexFunction;
+
+public class AddOneFunction implements VortexFunction<Integer, Integer> {
```
Missing comment.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addone/AddOneTest.java:
```diff
@@ -0,0 +1,51 @@
+import org.junit.Before;
+import org.junit.Test;
+
+public class AddOneTest {
```
Same here.

bgchun commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addone/AddOneTestUserCode.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Vector;
+import java.util.concurrent.ExecutionException;
+
+public class AddOneTestUserCode implements VortexUserCode {
```
Same here.

bgchun commented:
@johnyangk Excellent start! 
I'm done with my pass. Please take a look and leave a note on how you will address at a high level (you don't need to respond to the minor comments if you can just address them). Let me know when you're done with your pass.

markusweimer commented:
@johnyangk Please ping me when ready for another round of reviews. I'd like to understand this project a little better.

markusweimer commented:
I did a quick parse. One high-level bit that stood out is the liberal use of `public`. Given that we promise not to break `public` APIs in between REEF versions, we should be very careful with it. For instance, I don't think the vortex driver needs to be `public`, just its configuration.

markusweimer commented:
@sears You might find this project interesting, and a bit of a deja-vu :-)

johnyangk commented:
@bgchun Thank you for the thorough review! I will start addressing them now. For those that might be better to be addressed separately later, I'll create new JIRA sub-tasks for them.
@markusweimer Will do! And thank you for the comments.

bgchun commented:
@markusweimer Thanks!
@johnyangk Let me know when you're done with the pass with @markusweimer.

johnyangk commented:
I am still working on addressing @bgchun's comments. I will upload the code to here at once (probably sometime next week) when I'm ready and the discussion settles down.

I also created some JIRA sub-tasks under https://issues.apache.org/jira/browse/REEF-364. Please feel free to comment on them. Thank you!

johnyangk commented:
@bgchun I pushed the code to address your comments.
@markusweimer I am ready for the review.

It seems that recent change in `AllTestsSuite` on the master branch is causing the conflict with this PR, but I'm not sure how to safely resolve this conflict without accidentally wiping out our discussion here on this page. I'll do more research on this.

johnyangk commented:
@markusweimer Regarding your comment on `public`, I removed it from as many classes I can and for those I couldn't(because they were accessed across packages) I annotated `Unstable` as I expect there will be frequent changes.

johnyangk commented:
Resolved the conflict.

bgchun commented:
@markusweimer, could you do code review since you expressed doing it? After your pass, I will do my pass. We're hoping to include this initial version in the upcoming release. Thanks!

markusweimer commented:
@bgchun Thanks. I will do a pass, but please don't feel blocked by me. Go ahead and merge when you feel it is ready.

markusweimer commented on lang/java/reef-applications/reef-vortex/pom.xml:
```diff
@@ -20,7 +20,7 @@ under the License.
     <modelVersion>4.0.0</modelVersion>
-    <artifactId>reef-vortex</artifactId>
+    <artifactId>vortex</artifactId>
     <name>REEF Vortex</name>
```
Maybe also change the name to Vortex. However, I am sure that the ASF would want to call it "Apache REEF Vortex (incubating)" for now :-)

johnyangk replied:
Okay, then I'll just leave it as it is for now, since Tang/Wake are also like this.

markusweimer commented on lang/java/reef-applications/reef-vortex/pom.xml:
```diff
@@ -28,4 +28,76 @@ under the License.
+                    </filters>
+                </configuration>
+            </plugin>
+            <plugin>
```
This *should* be inherited from the parent POM, I believe. @tcNickolas should know for sure.

tcNickolas replied:
Yes, maven-checkstyle-plugin configuration is present only in the root POM, and it is inherited for all subprojects, so there no need to duplicate it here.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFuture.java:
```diff
@@ -0,0 +1,97 @@
+ * TODO[REEF-505]: Callback features for VortexFuture.
+ */
+@Unstable
+public class VortexFuture<TOutput> implements Future {
```
This is very similar to the original Task design in REEF, where `.submit()` would return a `TaskFuture<T> implements Future<RunningTask<T>>`. We decided to use an event driven approach instead, as the `Future` approach resulted in unwieldy client code involving lots of polling. Have you considered that alternative approach?

markusweimer replied:
Also, `Future` has a type parameter.

johnyangk replied:
Yes this is something @bgchun also pointed out I believe. I created a JIRA issue for it: https://issues.apache.org/jira/browse/REEF-505

Thanks for the catch, parameterized to `Future<TOutput>`.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexThreadPool.java:
```diff
@@ -0,0 +1,43 @@
+    this.vortexMaster = vortexMaster;
+  }
+
+  public <TInput extends Serializable, TOutput extends Serializable> VortexFuture<TOutput>
```
Please add JavaDoc.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/ExecuteTasklet.java:
```diff
@@ -0,0 +1,55 @@
+ * Request to execute a tasklet.
+ */
+@Unstable
+public class ExecuteTasklet<TInput extends Serializable, TOutput extends Serializable> implements VortexRequest {
```
Verbs as class names can be confusing. How about `TaskletExecutionRequest`?

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/ExecuteTasklet.java:
```diff
@@ -0,0 +1,55 @@
+  private final VortexFunction<TInput, TOutput> userFunction;
+  private final TInput input;
+
+  public ExecuteTasklet(final int taskletId,
```
Please add JavaDoc.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/TaskletException.java:
```diff
@@ -0,0 +1,48 @@
+ * Report of a tasklet exception.
+ */
+@Unstable
+public class TaskletException implements WorkerReport {
```
This is potentially confusing, as the class ends in `Exception` but does not, in fact, implement `Exception`. Maybe `TaskletFailure` instead?

johnyangk replied:
Agreed. Changed it to `TaskletFailureReport` with `Report` at the end as it implements `WorkerReport`.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/VortexRequest.java:
```diff
@@ -0,0 +1,38 @@
+    ExecuteTasklet
+  }
+
+  RequestType getType();
```
Please add JavaDoc

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/WorkerReport.java:
```diff
@@ -0,0 +1,39 @@
+    TaskletException
+  }
+
+  WorkerReportType getType();
```
Please add JavaDoc.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ * under the License.
+ */
+/**
+ * Commonly used by Driver and Evaluator.
```
Please be more specific. This is the common code for the Vortex Driver and Evaluator, correct?

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java:
```diff
@@ -0,0 +1,116 @@
+
+/**
+ * Default implementation of VortexMaster.
+ * This class is concurrently accessed by multiple threads and concurrency control is all done inside
```
Please consider using the jcip annotations for a more explicit documentation of the thread safety issues in this class. It is not only helping readers of the code, but some IDEs (IntelliJ IDEA, for instance) can validate them as well.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java:
```diff
@@ -0,0 +1,116 @@
+  private final ExecutorService schedulerThread = Executors.newSingleThreadExecutor();
+
+  @Inject
+  public DefaultVortexMaster(final RunningWorkers runningWorkers) {
```
If this is always injected, consider making the constructor `private`.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/VortexWorkerConf.java:
```diff
@@ -0,0 +1,47 @@
+ *  Vortex Worker configurations.
+ */
+@Unstable
+@TaskSide
```
Is this used on the Task side, or does it configure things for the Task side? If it is the latter, it is `@DriverSide`, to be consistent with the use of these annotations in REEF.

markusweimer commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/VortexWorkerConf.java:
```diff
@@ -0,0 +1,47 @@
+  public final class NumOfThreads implements Name<Integer> {
+  }
+
+  public static final RequiredParameter<Integer> NUM_OF_THREADS = new RequiredParameter<>();
```
Document.

markusweimer commented:
I've done a pass. Note that I have *not* looked at any of the implementation details, but focussed on APIs only. Some global comments on that:

  * Please consider using the JCIP annotations to document the thread safety concerns of this code.
  * Please add JavaDoc to all `public` methods and classes. At the very least, add them to the methods in all `public` interfaces.
  * Please consider the package-level javadoc as an opportunity to give the user and reader of the code guidance as to what the packages are for, in addition to what they contain.
  * Tang can instantiate an object based on a `private` constructor. Hence, making constructors `private` affords you greater freedom for future code changes. Consider doing that for all constructors which have an `@Inject` annotation.
  * Some package-private classes have `public` constructors. That seems unnecessary.
  * The evaluator restart logic is dangerous. If you suffer from a bug where *every* evaluator crashes upon launch, it will request new evaluators and launch them, just to see them fail. Consider adding a counter for the max number of failed evaluators.
  * Vortex relies on `Serializable` objects for the user code. This is very attractive from a usability perspective (exhibit A: Spark), but can also lead to enormous complications as those objects can refer to lots of local state (exhibit B: Spark's closure cleaner). For REEF, we erred on the side of making the latter impossible via Tang. I understand that this isn't the goal for Vortex. However, it would be good to have an idea of how to mitigate these issues before lots of code is written against the current API.

bgchun commented:
@markusweimer Thanks! Can we make the JavaDoc rule as part of checkstyle?

@johnyangk Let me know after you address Markus's comments. I will do a pass.

tcNickolas commented:
@bgchun Check for missing Javadoc will be enforced as checkstyle, but we cannot do it right now because of all violations we currently have. We need to fix those first. I've already created a bunch of JIRAs for this.

markusweimer commented:
@tcNickolas Can we enable that checkstyle check on a per-submodule basis? If so, we could enable it for vortex to make sure it is on a good path from the get-go.

johnyangk commented:
@markusweimer Thank you for the detailed comments! I will start working on them.
@bgchun Sure, I'll let you know when I'm done.

johnyangk commented:
@markusweimer I've addressed your comments. It would be good to double check using the checkstyle though.

Yes, we're still in the stage of figuring out the best way to ship code/data to VortexWorkers. Let's discuss together on [the JIRA issue](https://issues.apache.org/jira/browse/REEF-504). I agree that we should be careful not to build any serious Vortex applications until this issue is resolved.

@bgchun I'm ready. Please go ahead and review it.

bgchun commented:
@johnyangk Looks good. I left minor comments. Once I hear back from you, I will test and merge the code.

johnyangk commented:
@bgchun 
```java
VortexFuture enqueueMockedTasklet(final Tasklet tasklet)
```
Without this method currently there's no way for us to check the state of `Tasklet`. In current unit tests that test `DefaultVortexMaster`, we pass mocked `Tasklet`s whose ids we know to here and use their ids to track their state.(e.g. whether they were scheduled to a worker).

If we are to remove this method, I think we need to have each `VortexFuture` hold onto its corresponding `Tasklet`'s id. In this way we can use the following same method that the Vortex user api uses and still obtain the `Tasklet`'s id. But I do not prefer this approach because it exposes `Tasklet` ids to Vortex users.
```java
VortexFuture<TOutput> enqueueTasklet(final VortexFunction<TInput, TOutput> function, final TInput input)
```

Another approach is to change the structure of `DefaultVortexMasterTest` unit test.

Maybe I'm missing something? What are your thoughts?

bgchun commented:
@johnyangk If you need to change the behavior of `pendingTasklets` for your unit tests, maybe the dependency should be explicit for testing. Having a special method for unit tests indicates code smell.

johnyangk commented:
@bgchun Okay then I'll work on changing the overall code structure such that we can do without the special method.

johnyangk commented:
@bgchun 
Addressed your comments by
* Refactoring out the scheduler thread into a wake stage to be invoked by `VortexDriver`
* Creating injectable `PendingTasklets` to be injected into the scheduler and `DefaultVortexMaster`

By taking out these two dependencies from `DefaultVortexMaster`, the unit tests can now do the scheduling by itself without having to rely on special methods.

FYI: Fastest way to test Vortex
* At the REEF root directory, `mvn clean install -DskipTests`
* At lang\java\reef-applications, `mvn test`
* At lang\java\reef-tests, `mvn -Dtest=AddOneTest test`

bgchun commented:
@johnyangk Thanks. Looks good. I will test it and merge it if everything goes well.

johnyangk commented:
@bgchun 
Sorry, I just pushed a one-line change to fix a bug I just found. Please use that as well. Thank you!
